Add in-depth docs about how hilbert is written and works internally (#967)
netket/netket: docs/docs/hilbert.md

Those two abstract types are very different: `ContinuousHilbert` spaces are still experimental and we don't support yet many ways to manipulate them, while `DiscreteHilbert` spaces are much more developed and offer many utilities and handy functionalities.

Among the discrete Hilbert spaces the most important ones are {ref}`netket.hilbert.HomogeneousHilbert` spaces, where the local degrees of freedom are identical among different sites, of which there exist the {ref}`netket.hilbert.Fock`, {ref}`netket.hilbert.Spin` and {ref}`netket.hilbert.Qubit` implementation.

{ref}`netket.hilbert.TensorHilbert` represents tensor products of different homogeneous hilbert spaces, therefore it is not homogeneous. You can use it to represent composite systems such as spin-boson setups.

{ref}`netket.hilbert.DoubledHilbert` represents a space doubled through Choi's Isomorphism, which is where the density matrix of a physical system lives. It is used to work with dissipative/open systems.

## The AbstractHilbert interface

As we mentioned before, an Hilbert object represents at the same time a choice of Hilbert space and computational basis. 

> The reason why we need to specify an Hilbert space is evident, I hope.

Reviewer: I'd still want to reformulate this part. I can make a suggestion tomorrow, or you can do it if you want.

> The reason why we need to specify an Hilbert space is evident, I hope.

Author: Ah, i had forgotten to fix this. sorry

In practice all those operations rely on converting elements of the basis such as `[0,1,1,0]` to an integer index labelling all basis elements.

For the following examples, we will be using the {ref}`netket.hilbert.Qubit` hilbert space, whose local basis is `[0,1]`.

```python
>>> import netket as nk
>>> hi = nk.hilbert.Qubit(3)
Qubit(N=3)
```

Converting indices to basis elements can be performed through the {meth}`~netket.hilbert.DiscreteHilbert.numbers_to_states` method. 
When converting indices to a basis-element, NetKet relies on a sort of big-endian (or Most-Significant-Bit first) N-ary-encoding: for qubits, index $0$ will correspond to $|0,0,0\rangle$, index $1$ to $|0,0,1\rangle$, index $2$ to $|0,1,0\rangle$ and so on.
For hilbert spaces with larger local dimensions, all the local states are iterated continuously.

```python
>>> hi.numbers_to_states(0)
array([0., 0., 0.])
>>> hi.numbers_to_states(1)
array([0., 0., 1.])
>>> hi.numbers_to_states(2)
array([0., 1., 0.])
>>> hi.numbers_to_states(3)
array([0., 1., 1.])
>>> hi.numbers_to_states(7)
array([1., 1., 1.])
```

> It is also possible to perform the opposite transformation, and go from a basis element to an integer index using the {meth}`~netket.hilbert.DiscreteHilbert.states_to_numbers` method.

Reviewer: ```suggestion
It is also possible to perform the opposite transformation and go from a basis element to an integer index using the {meth}`~netket.hilbert.DiscreteHilbert.states_to_numbers` method.
```

```python
>>> hi.states_to_numbers(np.array([0,0,0]))
0
>>> hi.states_to_numbers(np.array([0,0,1]))
1
>>> hi.states_to_numbers(np.array([1,0,1]))
5
```

> Do notice that all those methods work with arrays too, and will convert an array of $M$ indices to a batch of states, that is, a matrix of size $M \times N$.

Reviewer: ```suggestion
Do notice that all those methods work with arrays too and will convert an array of $M$ indices to a batch of states, that is, a matrix of size $M \times N$.
```

(Hilbert)=
# The Hilbert module

```{eval-rst}
.. currentmodule:: netket.hilbert
```

> The [Hilbert](`netket.hilbert`) module defines the abstract Hilbert space API and some concrete implementations, such as {ref}`netket.hilbert.Spin`, {ref}`netket.hilbert.Fock` or others.

Reviewer: "such as ... and others" is somewhat redundant
```suggestion
The [Hilbert](`netket.hilbert`) module defines the abstract Hilbert space API and some concrete implementations, such as {ref}`netket.hilbert.Spin`, {ref}`netket.hilbert.Fock`.
```

An `Hilbert` object represents a Hilbert space together with a particular choice of computational basis.
They are needed to construct most other objects in NetKet, but they can also be useful to experiment and validate variational ansätze.

Hilbert space objects are all sub-classes of the abstract class {ref}`netket.hilbert.AbstractHilbert`, which defines the general API respected by all implementations. 

> You can see a birds-eye of the inheritance diagram among the various kinds of Hilbert spaces included with NetKet below (you can click on the nodes in the graph to go to their API documentation page). 

Reviewer: ```suggestion
You can see a birds-eye view of the inheritance diagram among the various kinds of Hilbert spaces included with NetKet below (you can click on the nodes in the graph to go to their API documentation page). 
```

Classes whose edge is dashed are abstract classes, while the others are concrete and can be instantiated.

```{eval-rst}
.. inheritance-diagram:: netket.hilbert
	:top-classes: netket.hilbert.AbstractHilbert
	:parts: 1

```

> {ref}`netket.hilbert.AbstractHilbert` makes very few assumptions on the structure of the resulting space and you will generally very rarely interact with it directly.
> 
> There are then two more abstract Hilbert space types: {ref}`netket.hilbert.DiscreteHilbert`, representing Hilbert spaces where the local degrees of freedom are countable, and {ref}`netket.hilbert.ContinuousHilbert`, representing the Hilbert spaces with continuous bases, such as particles in a box. 

Reviewer: ```suggestion
{ref}`netket.hilbert.AbstractHilbert` makes very few assumptions on the structure of the resulting space and you will generally very rarely interact with it directly.
Derived from `AbstractHilbert` are two less generic, but still abstract, types: {ref}`netket.hilbert.DiscreteHilbert`, representing Hilbert spaces where the local degrees of freedom are countable, and {ref}`netket.hilbert.ContinuousHilbert`, representing the Hilbert spaces with continuous bases, such as particles in a box.
```

> Those two abstract types are very different: `ContinuousHilbert` spaces are still experimental and we don't support yet many ways to manipulate them, while `DiscreteHilbert` spaces are much more developed and offer many utilities and handy functionalities.
> 
> The most important class of discrete Hilbert spaces are subclasses of {ref}`netket.hilbert.HomogeneousHilbert`, which is a space where the local degrees of freedom are identical among different sites. These subclasses are {ref}`netket.hilbert.Fock`, {ref}`netket.hilbert.Spin`, and {ref}`netket.hilbert.Qubit`.

Reviewer: ```suggestion
So far, the majority of NetKet development has focused `DiscreteHilbert` spaces which therefore have a much more developed API, while `ContinuousHilbert` is still experimental and does not yet support many operations.

The most important class of discrete Hilbert spaces are subclasses of {ref}`netket.hilbert.HomogeneousHilbert`, which is a tensor product of a finite number of local Hilbert spaces of the same kind, each with the same number of local degrees of freedom.
`HomogeneousHilbert` has the concrete subclasses {ref}`netket.hilbert.Fock`, {ref}`netket.hilbert.Spin`, and {ref}`netket.hilbert.Qubit`.
```

{ref}`netket.hilbert.TensorHilbert` represents tensor products of different homogeneous hilbert spaces, therefore it is not homogeneous. You can use it to represent composite systems such as spin-boson setups.

> {ref}`netket.hilbert.DoubledHilbert` represents a space doubled through [Choi's Isomorphism](https://en.wikipedia.org/wiki/Choi%E2%80%93Jamio%C5%82kowski_isomorphism).

Reviewer: Somehow that Wikipedia link (which involves trace-preserving complete positive maps) confuses me more: Is `DoubledHilbert` just ℋ ⊗ ℋ or is there more to it?

Author: it is.

In this example we show how to use it to build a space that behaves like {ref}`Fock`, while enforcing even parity.

```python
>>> import numba
>>>
>>> @numba.njit
>>> def accept_even(states):
>>> 	return states.sum(axis=-1) % 2 == 0
>>>
>>> n_max = 3; N_sites = 5
>>> hi = netket.hilbert.CustomHilbert(local_states=range(n_max), N=N_sites, constraint_fn=accept_even)
>>> hi.all_states()
array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 2.],
       [0., 0., 0., 1., 1.],
       [0., 0., 0., 2., 0.],
       ...
```

The constraint function sums the basis number (a number in `range(n_max)`) and then checks if it is even. 
Pleaes notice how we used `@numba.njit` to speed up the constraint.

If you then want to sample this space, you'll encounter the following error:

```python
>>> import jax
>>> hi.random_state(jax.random.PRNGKey(3), 3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/filippovicentini/Dropbox/Ricerca/Codes/Python/netket/netket/hilbert/abstract_hilbert.py", line 84, in random_state
    return random.random_state(self, key, size, dtype=dtype)
  File "plum/function.py", line 537, in plum.function.Function.__call__

>   File "/home/filippovicentini/Dropbox/Ricerca/Codes/Python/netket/netket/hilbert/random/custom.py", line 25, in random_state

Reviewer 2: Dropbox/Ricerca

Author: You don't like my folder structure?

Author: changed
